### PR TITLE
MINOR: Connect Javadocs improvements

### PR DESCRIPTION
**Pull request by yashmayya**
Some common themes:

- Replace blank lines (which are ignored) with `<p>` to separate paragraphs. For those wondering about the omitted closing `</p>` tags, this is actually the recommended style (see [here](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html)).
- Add Javadocs to some (but not all) classes that were missing them.
- Fix typos and grammatical issues.
- Add `{@link}` to code references wherever applicable.
- Replace manually numbered lists with `<ol>`.

For instance, the `AbstractHerder`'s rendered Javadoc before:
Screenshot 2022-12-03 at 8 01 08 PM

After:
Screenshot 2022-12-03 at 7 55 44 PM

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java [hunk 1]
```diff
@@ -131,8 +131,8 @@ public void reconfigure(Map<String, String> props) {
     /**
      * Validate the connector configuration values against configuration definitions.
      * @param connectorConfigs the provided configuration values
-     * @return List of Config, each Config contains the updated configuration information given
-     * the current configuration values.
+     * @return {@link Config}, essentially a list of {@link ConfigValue}s containing the updated configuration
+     * information given the current configuration values.
```
Comment (gharris1727) [inline connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java]:
```suggestion
     * @return a parsed and validated {@link Config} containing any relevant validation errors with the raw
     * {@code connectorConfigs} which should prevent this configuration from being used.
```

File: connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigOverridePolicy.java [hunk 2]
```diff
@@ -23,25 +23,25 @@
...
 public interface ConnectorClientConfigOverridePolicy extends Configurable, AutoCloseable {
 
 
     /**
-     * Worker will invoke this while constructing the producer for the SourceConnectors,  DLQ for SinkConnectors and the consumer for the
-     * SinkConnectors to validate if all of the overridden client configurations are allowed per the
-     * policy implementation. This would also be invoked during the validation of connector configs via the Rest API.
-     *
+     * Workers will invoke this while constructing producer for SourceConnectors, DLQs for SinkConnectors and
+     * consumers for SinkConnectors to validate if all of the overridden client configurations are allowed per the
```
Comment (gharris1727) [inline connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigOverridePolicy.java]: We probably don't need to enumerate all of the use-cases for kafka clients here, and can keep that scoped to the ConnectorClientConfigRequest javadoc.
```suggestion
     * Workers will invoke this before configuring per-connector Kafka admin, producer, and consumer
     * client instances to validate if all of the overridden client configurations are allowed per the
```

File: connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigRequest.java [hunk 3]
```diff
@@ -44,25 +44,25 @@ public ConnectorClientConfigRequest(
...
 
     /**
-     * Provides Config with prefix {@code producer.override.} for {@link ConnectorType#SOURCE}.
-     * Provides Config with prefix {@code consumer.override.} for {@link ConnectorType#SINK}.
-     * Provides Config with prefix {@code producer.override.} for {@link ConnectorType#SINK} for DLQ.
-     * Provides Config with prefix {@code admin.override.} for {@link ConnectorType#SINK} for DLQ.
+     * <p>Provides Config with prefix "{@code producer.override.}" for {@link ConnectorType#SOURCE}.
+     * <p>Provides Config with prefix "{@code consumer.override.}" for {@link ConnectorType#SINK}.
+     * <p>Provides Config with prefix "{@code producer.override.}" for {@link ConnectorType#SINK} for DLQ.
+     * <p>Provides Config with prefix "{@code admin.override.}" for {@link ConnectorType#SINK} for DLQ.
```
Comment (gharris1727) [inline connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigRequest.java]: We should also mention some more recent usages of this that haven't been updated in this documentation:
* Admin for fencing zombies in EOS source
* Admin for creating connector-specific source offset topics
* Consumer for connector-specific source offset topics
* Producer for connector-specific source offset topics

File: connect/api/src/main/java/org/apache/kafka/connect/sink/SinkTask.java [hunk 4]
```diff
@@ -88,15 +88,15 @@ public void initialize(SinkTaskContext context) {
     public abstract void start(Map<String, String> props);
 
     /**
-     * Put the records in the sink. Usually this should send the records to the sink asynchronously
-     * and immediately return.
-     *
+     * Put the records in the sink. This should either write them to the downstream system or batch them for
+     * later writing
```
Comment (gharris1727) [inline connect/api/src/main/java/org/apache/kafka/connect/sink/SinkTask.java]: I agree that it's not necessary for this javadoc to prescribe asynchronous behavior, but it should certainly point out the pitfall that asynchronous users need to take special care.
```suggestion
     * Put the records in the sink. If this method returns before the records are durably written,
     * the task must implement {@link #flush(Map)} or {@link #preCommit(Map)} to ensure that
     * only durably written record offsets are committed, and that no records are dropped during failures.
```

**Review by gharris1727 — COMMENTED** [review]
These really go unnoticed most of the time, but some of the mistakes are pretty silly now that you point them out. I'm glad that you're fixing them.

I just had a few suggestions for re-words and updates to stale info.

Thanks!

File: connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigOverridePolicy.java (hunk 2 above)
Comment (yashmayya) [reply connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigOverridePolicy.java]: Makes sense

File: connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigRequest.java (hunk 3 above)
Comment (yashmayya) [reply connect/api/src/main/java/org/apache/kafka/connect/connector/policy/ConnectorClientConfigRequest.java]: Thanks, that's a great call out!

**Review by yashmayya — COMMENTED** [review]
Thanks for the review @gharris1727! I've addressed your comments.

**Review by gharris1727 — APPROVED** [review]

**Review by mimaison — APPROVED** [review]
LGTM!

**Comment by yashmayya** [comment]
Thanks Mickael!

**Pull request merged**